selenium/src/main/java/com/deque/html/axecore/selenium/AxeBuilder.java

```diff
@@ -686,32 +702,53 @@ private ArrayList<String> runPartialRecursive(
       String resResponse =
           (String)
               WebDriverInjectorExtensions.executeAsyncScript(
-                  webDriver, runPartialScript, context, options);
-      ArrayList<String> partialResults = new ArrayList<String>();
+                  webDriver, runPartialScript, context, options, frameStack);
       partialResults.add(resResponse);
       if (disableIframeTesting) {
         return partialResults;
       }
 
       for (FrameContext fc : contexts) {
-        Object frameContext = AxeReporter.serialize(fc.getFrameContext());
-        Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
-        Object frame =
-            WebDriverInjectorExtensions.executeScript(webDriver, shadowSelectScript, frameSelector);
-
-        if (frame instanceof String) {
-          webDriver.switchTo().frame((String) frame);
-        } else if (frame instanceof WebElement) {
-          webDriver.switchTo().frame((WebElement) frame);
-        } else if (frame instanceof Integer) {
-          webDriver.switchTo().frame((Integer) frame);
-        } else {
+        try {
+          Object frameContext = AxeReporter.serialize(fc.getFrameContext());
+          Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
+          frameStack.push(frameSelector);
+          Object frame =
+              WebDriverInjectorExtensions.executeScript(
+                  webDriver, shadowSelectScript, frameSelector);
+
+          if (frame instanceof String) {
+            webDriver.switchTo().frame((String) frame);
+          } else if (frame instanceof WebElement) {
+            webDriver.switchTo().frame((WebElement) frame);
+          } else if (frame instanceof Integer) {
+            webDriver.switchTo().frame((Integer) frame);

```

Comment:
Why are these casts necessary? Can we not just do:

```suggestion
          if ((frame instanceof String) || (frame instanceof WebElement) || (frame instanceof Integer)) {
            webDriver.switchTo().frame(frame);
```

Reply:
Java uses static method dispatch based on type. While the method is overloaded for `String`, `WebElement`, and int, it must be resolved to a single one of those overloads at compile time.
All that to say, we have to have the individual type casts because Java just looks to see if it can call `.frame(Object)` and fails to compile.